### PR TITLE
Adds some new features to NTSL

### PR DESCRIPTION
**Pull request by monster860**
Alllllright. I've got some fancy new features for NTSL here.

The first of them is that the tcomms server can now interface with machines. Right now, it can only interface with the crew monitoring computer and the new "server interface" computer. You can currently link a machine to a server by using the tcomms traffic control console. The machine has to be in the same area, and it has to have the proper callbacks for tcomms interfacing.

The new server interface is pretty cool. It allows you to send a custom UI to it, which will be displayed.

Screenshots:

:cl: monster860
rscadd: Adds a new NTSL scripting feature, where you control machines.
rscadd: Adds crew manifest access to NTSL
rscadd: Adds PDA messaging access to NTSL
rscadd: New server interface console: A machine-programmed console. Made in R&D, called "server interface"
/:cl:

### CODE REVIEW

File: paradise.dme [hunk 1]
```diff
@@ -10,7 +10,7 @@
 #define DEBUG
 // END_PREFERENCES
 // BEGIN_INCLUDE
-#include "_maps\cyberiad.dm"
+#include "_maps\test.dmm"
```
Comment (tigercat2000) [inline paradise.dme]: forgot to untick this

File: code/game/machinery/telecomms/telecomunications.dm [hunk 2]
```diff
@@ -465,6 +465,7 @@ var/global/list/obj/machinery/telecomms/telecomms_list = list()
 	var/rawcode = ""	// the code to compile (raw text)
 	var/datum/TCS_Compiler/Compiler	// the compiler that compiles and runs the code
 	var/autoruncode = 0		// 1 if the code is set to run every time a signal is picked up
+	var/
```
Comment (tigercat2000) [inline code/game/machinery/telecomms/telecomunications.dm]: err

**Comment by FlattestGuitar** [comment]
I feel like making additions to the code that you're only going to use yourself is an new and exciting form of powergaming. What a world we live in.

:+1: x :100: for lynch lists.

**Comment by fludd12** [comment]
If you could somehow separate this from telecomms and instead make a machine scripting console

I'd probably marry you or something.

**Comment by monster860** [comment]
I'd rather keep it linked with telecomms. As it is, it can make broadcasts and shit.

**Comment by monster860** [comment]
> :+1: x :100: for lynch lists

That lynch list script I made was not well written. If I wanted to do it properly, I would have copied in names that appeared on the radio traffic and used those.

**Comment by Fox-McCloud** [comment]
I'd rather not expand the functionality of a largely unmaintained feature that causes performance issues, to be honest. Least of all NTSL.

**Comment by tigercat2000** [comment]
@Fox-McCloud it's not "maintained" because there's nothing particularly wrong with it

**Comment by ProperPants** [comment]
Honestly, this is pretty fine, but @monster860 should make a guide for TComms if this goes through.

**Comment by PJB3005** [comment]
@tigercat2000 having fucked around with NTSL I can tell you there's a lot wrong with it, even if it's just the language itself and not the implementation, but even then I don't think that should be a showstopper for something like this because :+1: :x: :100:

**Comment by monster860** [comment]
@PJB3005 Yep. Im really annoyed the proc for adding shit to the list is called push_back. Doesnt make sense. Also the no function in function thing. Finally sleep() causes weird bugs.

**Comment by FlattestGuitar** [comment]
@monster860 , `push_back` exactly what you use in C++ for vectors, so I guess that's where it comes from.

Also, muh sprites.

**Comment by monster860** [comment]
If you have created these sprites then ... send em in when school gets out for me.

**Comment by monster860** [comment]
Woo, doors can be controlled. Next to add atmos machines control and be able to suck the air out of a room and replace it with plasma. What a fun time.

**Comment by PJB3005** [comment]
AAC interfacing when?

**Comment by monster860** [comment]
You mean APC? Soon.

**Comment by PJB3005** [comment]
No, AAC. Atmospheric Automations Console.

https://github.com/ParadiseSS13/Paradise/tree/master/code/modules/atmos_automation

Programming atmos equipment.

**Comment by monster860** [comment]
Ah, of course. What does it do?

**Comment by optimumtact** [comment]
Just wait until your players figure out you can lock up the server with a well coded NTSL script

File: code/game/machinery/telecomms/server_interface.dm [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	var/list/ui_elements = list("text~Welcome. The telecomms server has not been configured to display a UI on this server interface. Press init to send \"init\" signal to server.","button~Init~init")
+
+/obj/machinery/computer/telecomms/server_interface/attack_hand(mob/user)
+	if(..())
+		return
+	interact(user)
+
+/obj/machinery/computer/telecomms/server_interface/attackby(obj/item/W, mob/user, params)
+	if(istype(W,/obj/item/weapon/pen))
+		user << "<span class='notice'>You insert \the [W]'s tip into \the [src]'s reset hole.</span>"
```
Comment (tigercat2000) [inline code/game/machinery/telecomms/server_interface.dm]: `to_chat(user, "<span class='notice'>You insert \the [W]'s tip into \the [src]'s reset hole.</span>")`

File: code/game/machinery/telecomms/server_interface.dm [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+		if(do_after(user, 50, target=src))
+			user << "<span class='notice'>You reset \the [src].</span>"
```
Comment (tigercat2000) [inline code/game/machinery/telecomms/server_interface.dm]: to_chat

File: code/game/machinery/telecomms/server_interface.dm [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+			ui_elements = initial(ui_elements)
+		else
+			user << "<span class='warning'>You pull \the [W] out too early, and fail to reset it</span>"
```
Comment (tigercat2000) [inline code/game/machinery/telecomms/server_interface.dm]: to_chat

**Comment by Earthdivine** [comment]
I just noticed this exists.

I will have to say :-1: to this, imo. After seeing how abusing NTSL was in the past (Did you know you can set it up to operate entirely out of a borg module set?), this means this makes the station a playground to anyone who knows NTSL.

**Comment by monster860** [comment]
@Earthdivine its not as OP as you think. Currently you can only access shit in the same '/area' as the server. Im also gonna add a relay object, which is big and blinky and easy to notice.

**Comment by monster860** [comment]
Alright, coming soon with fun shit

**Comment by tigercat2000** [comment]
:-1:
Literally "make NTSL more abuseable"
>Hide relays in maint/closet/literally anywhere else
>win

**Comment by monster860** [comment]
> closet

How does one hide a relay in a closet?

> maint

Maint area is not necessarily the area you want to control

**Comment by tigercat2000** [comment]
@monster860 normal station areas can extend into maint in some places

**Comment by PJB3005** [comment]
So now shit mappers is an excuse for preventing this awesome PR?

**Comment by tigercat2000** [comment]
@PJB3005 I don't consider it so "awesome" anymore.
It massively increases the power of a little scripting language that can already crash the server. Unless the relays are machinery that actually has to be inside the area and powered, it's incredibly abusable- If they are machinery that has to be inside the area, it makes it practically pointless as a lot of the station is separated by area directly at the doors.

**Comment by monster860** [comment]
Plan is for relays to have to be inside the area and powered. The point is, that if you have a giant relay hanging out in **insert area name here**, people will know that shenanigans are going on with telecomms

**Comment by PJB3005** [comment]
> that can already crash the server.

The admins get notified pretty clearly if that happens.

It's the fucking AAC all over again. SS13 is what it is due to the amount of power and depth you have in what you can do. Even with this feature it'll barely get used.

**Comment by tigercat2000** [comment]
Okay, I know I'm a wishy-washy useless opinion-giver, but after re-considering this, I actually do.. kinda? Like it.
I'm imagining a hallway full of doors made by RCD with a relay in the middle of them, then using telecomms to lock it down, which would be hilarious.

File: code/game/machinery/doors/airlock.dm [hunk 6]
```diff
@@ -1066,4 +1067,86 @@ About the new airlock wires panel:
...
-	return
\ No newline at end of file
+	return
+
+/obj/machinery/door/airlock/server_interface(var/list/argslist)
+	if(argslist && argslist.len >= 2)
+		var/activate = argslist[2]
+		switch(argslist[1])
+			if("idscan")
+				if(isWireCut(AIRLOCK_WIRE_IDSCAN))
```
Comment (tigercat2000) [inline code/game/machinery/doors/airlock.dm]: don't use empty if statements for an else-if chain
do 
```
if(!isWireCut(AIRLOCK_WIRE_IDSCAN))
    if(activate && aiDisabledIdScanner)
			aiDisabledIdScanner = 0
	else if(!activate && !aiDisabledIdScanner)
			aiDisabledIdScanner = 1
```
that goes for all of these

File: code/modules/scripting/Implementations/Telecomms.dm [hunk 7]
```diff
@@ -311,4 +332,39 @@
...
+		for (var/obj/machinery/message_server/MS in message_servers)
+			if(MS.active)
+				useMS = MS
+				break
+	if(useMS)
+		for(var/obj/item/device/pda/P in PDAs)
+			if (P.owner==recipient)
+				useMS.send_pda_message("[P.owner]", "[S.id] Automated Broadcast", message)
+				var/datum/data/pda/app/messenger/PM = P.find_program(/datum/data/pda/app/messenger)
+				PM:notify("<b>Message from [S.id] (Automated Message), </b>\"[message]\" (<i>Unable to Reply</i>)", 0)
```
Comment (tigercat2000) [inline code/modules/scripting/Implementations/Telecomms.dm]: don't use colons

File: code/modules/scripting/Implementations/_Logic.dm [hunk 8]
```diff
@@ -262,6 +262,16 @@ proc/n_round(var/num)
...
 		return log(num)
 
+/proc/n_crew_manifest()
+	var/list/returned_list = list()
+	for(var/datum/data/record/R in data_core.general)
+		var/name = sanitize(R.fields["name"])
+		var/rank = sanitize(R.fields["rank"])
+		var/real_rank = R.fields["real_rank"]
+		var/isactive = R.fields["p_stat"]
+		returned_list[++returned_list.len] = list("name"=name,"rank"=rank,"real_rank"=real_rank,"active"=isactive);
```
Comment (tigercat2000) [inline code/modules/scripting/Implementations/_Logic.dm]: don't use semicolons either

Comment (Spacemanspark) [inline code/modules/scripting/Implementations/_Logic.dm]: [INSERTRAGEABOUTHOWICAN'TUSESEMICOLONSHERE]

**Comment by monster860** [comment]
> I'm imagining a hallway full of doors made by RCD with a relay in the middle of them, then using telecomms to lock it down, which would be hilarious.

That's a lot of doors to link

**Comment by monster860** [comment]
At this point, I believe, all that needs doing is MOAR CALLBACKS.

**Comment by monster860** [comment]
Morgue conflicts fixed.

Anyways, I would like to request review. At this point, this is pretty much complete. Besides adding more machines to the list of things it can control, there's nothing really to do.

You can merge if you want.

**Comment by ScratKnapp** [comment]
So what now? It's been 2 months afterall.

**Comment by monster860** [comment]
Morgue conflicts fixed.

**Comment by Fox-McCloud** [comment]
After taking this into consideration and discussing this over with @TheDZD 

we'd really rather not have telecomms receive such a buff--it's already massively powerful and accessible to only 3 crewmembers+the AI. It's not a system we believe is good for the long-term health of the game, nor one that has much potential for anything other than abusive mechanics.

**Comment by PJB3005** [comment]
Ah, the AAC paradox.

This is exactly what happened when I ported the AAC. 

You're overestimating how powerful this is. Just fucking merge it, if it really is OP, it can be reverted later.

**Comment by tigercat2000** [comment]
nothing is ever reverted tho

**Comment by PJB3005** [comment]
Literally 2 button presses.

**Comment by PJB3005** [comment]
Is there at least a forum thread to discuss this?

**Comment by Spacemanspark** [comment]
Nope.

**Pull request closed without merging**